server/debug/src/symbol.rs

```diff
@@ -0,0 +1,254 @@
+    pub fn line_from_address(&mut self, address: usize) -> io::Result<(OsString, u32, usize)> {
+        unsafe {
+            let mut displacement = 0;
+            let mut line = winapi::IMAGEHLP_LINEW64 {
+                SizeOfStruct: mem::size_of::<winapi::IMAGEHLP_LINEW64>() as winapi::DWORD,
+                ..mem::zeroed()
+            };
+
+            if dbghelp::SymGetLineFromAddrW64(
+                self.0, address as winapi::DWORD64, &mut displacement, &mut line
+            ) == winapi::FALSE {
+                return Err(io::Error::last_os_error());
+            }
+
+            let mut len = 0;
+            while *line.FileName.offset(len as isize) != 0 { len += 1; }
+
+            let name = slice::from_raw_parts(line.FileName, len);
+            let name = OsString::from_wide(name);
+
+            Ok((name, line.LineNumber, displacement as usize))
+        }
+    }
+
+    /// Iterate through the frames of a thread's stack
+    ///
+    /// The thread should be part of an attached child process which is currently paused to handle
+    /// a debug event.
+    pub fn walk_stack(&mut self, thread: winapi::HANDLE) -> io::Result<StackFrames> {
+        unsafe {
+            let mut context = winapi::CONTEXT {
+                ContextFlags: winapi::CONTEXT_FULL,
+                ..mem::zeroed()
+            };
+            if kernel32::GetThreadContext(thread, &mut context) == winapi::FALSE {
+                return Err(io::Error::last_os_error());
+            }
+
+            fn flat(address: winapi::DWORD64) -> winapi::ADDRESS64 {
+                winapi::ADDRESS64 { Offset: address, Mode: winapi::AddrModeFlat, Segment: 0 }
+            }
+            let stack = winapi::STACKFRAME64 {
+                AddrPC: flat(context.Rip),
+                AddrFrame: flat(context.Rbp),
+                AddrStack: flat(context.Rsp),
+                ..mem::zeroed()
+            };
+
+            Ok(StackFrames { process: self.0, thread, context, stack })
+        }
+    }
+
+    /// Enumerate the local symbols of a stack frame
+    ///
+    /// Addresses are base-pointer-relative.
+    pub fn enumerate_symbols<F>(&self, frame: &StackFrame, mut f: F) -> io::Result<()>
+        where F: FnMut(&Symbol, usize) -> bool
+    {
+        unsafe {
+            let mut stack_frame = winapi::IMAGEHLP_STACK_FRAME {
+                InstructionOffset: frame.stack.AddrPC.Offset,
+                ..mem::zeroed()
+            };
+            if dbghelp::SymSetContext(self.0, &mut stack_frame, ptr::null_mut()) == winapi::FALSE {
+                return Err(io::Error::last_os_error());
+            }
+
+            if dbghelp::SymEnumSymbolsW(
+                self.0, 0, ptr::null(), Some(enum_symbols::<F>), &mut f as *mut _ as *mut _
+            ) == winapi::FALSE {
+                return Err(io::Error::last_os_error());
+            }
+
+            return Ok(());
+        }
+
+        #[allow(non_snake_case)]
```

Comment:
Yes, though this syntax (`#[allow(non_snake_case)]`) applies to the following item. If you want it to apply to its containing module (as when putting it at the top of `serde_types.rs`) you want the syntax `#![allow(non_snake_case)]` (note the `!`).